File: ShootKit/Sample Projects/Objective-C Sample/App/Lib/CameraSource.m
//
//  CameraSource.m
//  Objective-C Sample
//
//  Created by Michael Forrest on 02/06/2023.
//

#import "CameraSource.h"
@import AVKit;

@interface CameraSource(){
    AVCaptureSession * session;
    AVCaptureVideoDataOutput* output;
    dispatch_queue_t queue;
}
@property (weak,nonatomic) id<AVCaptureVideoDataOutputSampleBufferDelegate> delegate;
@end


@implementation CameraSource
- (instancetype)initWithCaptureDelegate:(id<AVCaptureVideoDataOutputSampleBufferDelegate>)delegate{
    self = [super init];
    if (self) {
        self.delegate = delegate;
        [self build];
    }
    return self;
}

-(void)build{
    self.availableCameras = [AVCaptureDeviceDiscoverySession discoverySessionWithDeviceTypes:@[AVCaptureDeviceTypeBuiltInWideAngleCamera, AVCaptureDeviceTypeExternalUnknown] mediaType:AVMediaTypeVideo position:AVCaptureDevicePositionUnspecified].devices;
    
    session = [[AVCaptureSession alloc] init];
    output = [[AVCaptureVideoDataOutput alloc] init];
    queue = dispatch_queue_create("camera", DISPATCH_QUEUE_SERIAL);
    
    [session beginConfiguration];
    output.videoSettings = [[NSDictionary alloc] initWithObjectsAndKeys:@(kCVPixelFormatType_32BGRA),(NSString*)kCVPixelBufferPixelFormatTypeKey , nil];
    output.alwaysDiscardsLateVideoFrames = true;
    [output setSampleBufferDelegate:self.delegate queue:queue];
    [session addOutput:output];
    [session commitConfiguration];
    
    [self selectCamera: self.availableCameras.firstObject];
}
-(void)selectCamera:(AVCaptureDevice *)camera{
    self.selectedCamera = camera;
    
    NSError*error;
    AVCaptureDeviceInput * input = [[AVCaptureDeviceInput alloc] initWithDevice:camera error:&error];
    
    [session stopRunning];
    [session beginConfiguration];
    [session.inputs enumerateObjectsUsingBlock:^(__kindof AVCaptureInput * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [session removeInput:obj];
    }];
    [session addInput:input];
    [session commitConfiguration];
    [session startRunning];
}

@end

File: ShootKit/Sample Projects/Swift Sample/App/Video Pencil/CameraSource.swift
//
//  CameraSource.swift
//  ShootKit
//
//  Created by Michael Forrest on 02/06/2023.
//

import Foundation
import AVKit

class CameraSource: ObservableObject{
    var availableDevices: [AVCaptureDevice] = AVCaptureDevice.DiscoverySession(deviceTypes: [.builtInWideAngleCamera, .externalUnknown], mediaType: .video, position: .unspecified).devices
    
    @Published var selectedCamera: AVCaptureDevice?{
        didSet{
            if let device = selectedCamera{
                try? connectCamera(device: device)
            }
        }
    }
    
    let output = AVCaptureVideoDataOutput()
    let session = AVCaptureSession()
    let queue = DispatchQueue(label: "camera", qos: .userInteractive)
    
    init(captureDelegate: AVCaptureVideoDataOutputSampleBufferDelegate) {
        buildCaptureSession(captureDelegate: captureDelegate)
        selectedCamera = availableDevices.first
    }
    
    func buildCaptureSession(captureDelegate: AVCaptureVideoDataOutputSampleBufferDelegate){
        session.beginConfiguration()
        output.videoSettings = [
            kCVPixelBufferPixelFormatTypeKey as String: Int(kCVPixelFormatType_32BGRA)
        ]
        output.alwaysDiscardsLateVideoFrames = true
        output.setSampleBufferDelegate(captureDelegate, queue: queue)
        session.addOutput(output)
      
        session.commitConfiguration()
      
    }
    func connectCamera(device: AVCaptureDevice) throws{
        let input = try AVCaptureDeviceInput(device: device)
        session.stopRunning()
        session.beginConfiguration()
        for input in session.inputs{
            session.removeInput(input)
        }
        session.addInput(input)
        session.commitConfiguration()
        session.startRunning()
    }
}
